write_buffer/src/core.rs

```diff
@@ -26,3 +31,151 @@ pub trait WriteBufferReading: Sync + Send + std::fmt::Debug + 'static {
+
+        fn reading(&self) -> Self::Reading;
+    }
+
+    pub async fn perform_generic_tests<T>(adapter: T)
+    where
+        T: TestAdapter,
+    {
+        test_single_stream_io(&adapter).await;
+        test_multi_stream_io(&adapter).await;
+        test_multi_writer_multi_reader(&adapter).await;
+    }
+
+    async fn test_single_stream_io<T>(adapter: &T)
+    where
+        T: TestAdapter,
+    {
+        let context = adapter.new_context(1).await;
+
+        let entry_1 = lp_to_entry("upc user=1 100");
+        let entry_2 = lp_to_entry("upc user=2 200");
+        let entry_3 = lp_to_entry("upc user=3 300");
+
+        let writer = context.writing();
+        let reader = context.reading();
+
+        let mut stream = reader.stream();
+        let waker = futures::task::noop_waker();
+        let mut cx = futures::task::Context::from_waker(&waker);
+
+        // empty stream is pending
+        assert!(stream.poll_next_unpin(&mut cx).is_pending());
+
+        // adding content allows us to get results
+        writer.store_entry(&entry_1, 0).await.unwrap();
+        assert_eq!(stream.next().await.unwrap().unwrap().entry(), &entry_1);
+
+        // stream is pending again
+        assert!(stream.poll_next_unpin(&mut cx).is_pending());
+
+        // adding more data unblocks the stream
+        writer.store_entry(&entry_2, 0).await.unwrap();
+        writer.store_entry(&entry_3, 0).await.unwrap();
+        assert_eq!(stream.next().await.unwrap().unwrap().entry(), &entry_2);
+        assert_eq!(stream.next().await.unwrap().unwrap().entry(), &entry_3);
+
+        // stream is pending again
+        assert!(stream.poll_next_unpin(&mut cx).is_pending());
+    }
+
+    async fn test_multi_stream_io<T>(adapter: &T)
+    where
+        T: TestAdapter,
+    {
+        let context = adapter.new_context(1).await;
+
+        let entry_1 = lp_to_entry("upc user=1 100");
+        let entry_2 = lp_to_entry("upc user=2 200");
+        let entry_3 = lp_to_entry("upc user=3 300");
+
+        let writer = context.writing();
+        let reader = context.reading();
+
+        let mut stream_1 = reader.stream();
+        let mut stream_2 = reader.stream();
+        let waker = futures::task::noop_waker();
+        let mut cx = futures::task::Context::from_waker(&waker);
+
+        // empty streams is pending
+        assert!(stream_1.poll_next_unpin(&mut cx).is_pending());
+        assert!(stream_2.poll_next_unpin(&mut cx).is_pending());
+
+        // streams poll from same source
+        writer.store_entry(&entry_1, 0).await.unwrap();
+        writer.store_entry(&entry_2, 0).await.unwrap();
+        writer.store_entry(&entry_3, 0).await.unwrap();
+        assert_eq!(stream_1.next().await.unwrap().unwrap().entry(), &entry_1);
+        assert_eq!(stream_2.next().await.unwrap().unwrap().entry(), &entry_2);
+        assert_eq!(stream_1.next().await.unwrap().unwrap().entry(), &entry_3);
+
+        // stream1 is pending again
```

Comment:
Nit: this looks like it's testing that both streams are pending again, but the comment is emphasizing `stream1` only?

```suggestion
        // both streams are pending again
```

Reply:
copy+paste error